### PR TITLE
Add __spreadArrays helper

### LINKED ISSUE #8856: Spread operator is not correctly translated into JS
**TypeScript Version:**  1.8 
**Target:** ES5

The following code: 

``` ts
[...(new Array(5))]
```

translates into: 

``` ts
(new Array(5)).slice();
```

However, the ES6 meaning is not the same. See the output below: 

``` ts
> [...(new Array(5))]
[ undefined, undefined, undefined, undefined, undefined ]
> (new Array(5)).slice();
[ , , , ,  ]
```

Expected behavior: 

``` ts
Array.apply(null, Array(5))
```

### PR DESCRIPTION
**Pull request by rbuckton**
Adds a `__spreadArrays` helper for a more accurate spread behavior when *not* using `--downlevelIteration`.

Fixes #8856

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test

File: src/compiler/factory.ts [hunk 1]
```diff
@@ -3693,6 +3693,30 @@ namespace ts {
         );
     }
 
+    export const spreadArraysHelper: UnscopedEmitHelper = {
```
Comment (sheetalkamat) [inline src/compiler/factory.ts]: You would want to add it to `allUnscopedEmitHelpers`

**Comment by MaxGraey** [comment]
What about speedup this?
```js
var __spreadArrays = function () {
  for (var i = 0, len = 0; i < arguments.length; ++i) len += arguments[i].length;
  var arr = new Array(len);
  for (var i = 0, k = 0, l = arguments.length; i < l; ++i) {
    for (var j = 0, a = arguments[i], jlen = a.length; j < jlen; ++j, ++k)
      arr[k] = a[j];
  }
  return arr;
};
```
On V8 7.4 this faster by ~3x in average case.

File: src/compiler/transformers/es2015.ts [hunk 2]
```diff
@@ -3840,17 +3840,33 @@ namespace ts {
...
                 }
 
-                // Rewrite using the pattern <segment0>.concat(<segment1>, <segment2>, ...)
-                return createArrayConcat(segments.shift()!, segments);
+                return createSpreadArraysHelper(context, segments);
             }
         }
 
+        function isPackedElement(node: Expression) {
+            return !isOmittedExpression(node);
```
Comment (weswigham) [inline src/compiler/transformers/es2015.ts]:
```ts
const innerSpreadEmpty = [...[,,,]]
```
?

Comment (rbuckton) [reply src/compiler/transformers/es2015.ts]: Works just fine.

Comment (rbuckton) [reply src/compiler/transformers/es2015.ts]: It basically catches `[...[1, 2]]` and turns it into `[1, 2]` rather than calling the helper.

**Comment by rbuckton** [comment]
@typescript-bot test this

**Comment by rbuckton** [comment]
@typescript-bot test this

**Comment by rbuckton** [comment]
Verified that the only differences in our RWC suite are expected.

**Pull request merged**